[SYCLLowerIR] Add support for FPGA latency control extension

asudarsa opened the pull request:
This PR adds support to lower compile time properties and generate llvm pointer annotations ready to be consumed by SPIR-V gen.
The compile-time properties lowered are latency-anchor-id and latency-constraint.

Thanks

asudarsa commented:
Hi @steffenlarsen 

Can you please take a look?

Thanks

steffenlarsen reviewed (approved):
Looks good to me!

Only nit I have is that I would prefer the test to be a little more focused, that is it has a lot of unnecessary information which could be removed without affecting the test itself.

asudarsa commented on llvm/test/SYCLLowerIR/CompileTimePropertiesPass/sycl-latency-control.ll:
```diff
@@ -7,60 +7,57 @@ target triple = "spir64-unknown-unknown"
 
 $_ZTSZ4fooEUlvE_ = comdat any
 
-@.str = private unnamed_addr addrspace(1) constant [16 x i8] c"sycl-properties\00", section "llvm.metadata"
```
I removed addrspace(1) to make the test simpler.

asudarsa commented on llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:
```diff
@@ -41,6 +41,7 @@ constexpr uint32_t SPIRV_PIPELINE_ENABLE_DECOR = 5919;
 enum class DecorValueTy {
   uint32,
   boolean,
+  string,
```
Added a string DecorValueTy to enabling the store of a list of literals as a comma-separated string.

asudarsa commented:
> Looks good to me!
> 
> Only nit I have is that I would prefer the test to be a little more focused, that is it has a lot of unnecessary information which could be removed without affecting the test itself.

I modified the test to have opaque pointers and also cleaned it up a bit.

asudarsa commented:
@intel/dpcpp-tools-reviewers 

Can you please take a look? 

Thanks

steffenlarsen commented on llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:
```diff
@@ -491,9 +519,14 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
   const Value *IntrAnnotStringArg = IntrInst->getArgOperand(1);
-  if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
-    if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+  if (IntrAnnotStringArg->getType()->isOpaquePointerTy()) {
+    if (auto *C = dyn_cast<Constant>(IntrAnnotStringArg))
       AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  } else {
+    if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
+      if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+        AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  }
```
Nit;
```suggestion
  if (IntrAnnotStringArg->getType()->isOpaquePointerTy())
    AnnotStrArgGV = dyn_cast<GlobalVariable>(IntrAnnotStringArg);
  else if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
    AnnotStrArgGV = dyn_cast<GlobalVariable>(GEP->getOperand(0));
```

steffenlarsen reviewed (approved):
Looks good to me! just a few minor simplification nits

asudarsa commented:
Test failures are not related to this change and are found to be failing in other locations as well. Thanks

asudarsa commented:
Hi @sarnex 

Can you please help to take a look at this change and help to merge?

Thanks

sarnex commented on llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:
```diff
@@ -77,6 +78,26 @@ MDNode *buildSpirvDecorMetadata(LLVMContext &Ctx, uint32_t OpCode,
+  MD.push_back(ConstantAsMetadata::get(
+      Constant::getIntegerValue(Ty, APInt(32, OpCode))));
+  MD.push_back(
+      ConstantAsMetadata::get(ConstantDataArray::getString(Ctx, Value, true)));
```
nit: can we add a comment on what true represents, like
```
ConstantDataArray::getString(Ctx, Value, /*whatever=*/true)
```
